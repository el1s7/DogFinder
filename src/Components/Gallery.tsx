import { Alert, Loading } from 'elementz';
import { useEffect, useRef } from 'react';
import Masonry from 'react-masonry-css'

import API from '../Helpers/api';
import { useMultiState } from '../Helpers/Functions';
import GalleryItem from './GalleryItem';

interface GalleryState {
	images: string[],
	limit: number,
	scrollEnd: boolean,
	isWaterfall: boolean,
	isFetching: boolean
}


function Gallery(props: {
	breed: string[],
	empty: JSX.Element,
	// Add initial images for unit-tests
	initialImages?: string[]
}) {

	const [breed, subBreed] = props.breed;

	const [state, setState]= useMultiState<GalleryState>({
		images: props.initialImages || [],
		isFetching: false,
		isWaterfall: true,
		 // Fetch more items depending on screen width
		limit: Math.floor(window.innerWidth * 0.03),
		scrollEnd: true
	});

	const lastBreed = useRef(breed + subBreed);

	const handleScroll = () => {
		const scrollCurrentPosition = (document.documentElement.scrollTop + (document.documentElement.clientHeight * 1.7));
		const scrollTotalHeight = document.documentElement.scrollHeight
		const isScrollFinished = scrollCurrentPosition >= scrollTotalHeight;

		if (isScrollFinished) {
			setState({
				scrollEnd: true
			});
		}
	}

	// Add Scroll Listener
	useEffect(() => {
		window.addEventListener("scroll", handleScroll);

		return () => {
			window.removeEventListener("scroll", handleScroll);
		}
	}, []);

	// Reset Gallery on Breed Change
	useEffect(() => {
		if (!breed || (lastBreed.current !== (breed + subBreed))) {
			setState({
				images: [],
				scrollEnd: true,
			});
		}

		lastBreed.current = (breed + subBreed);
	}, [breed, lastBreed]);

	// Get Images
	useEffect(() => {
		if (breed && state.scrollEnd) {
			setState({
				isFetching: true
			});

			API.dogCeo.random({
				breed,
				limit: state.limit,
				subbreed: subBreed,
			})
				.then((res) => {
					setState((currentState) => ({
						images: [
							...currentState.images,
							...res.message
						]
					}));
				})
				.catch(() => {
					Alert.danger("Failed loading doggos ¯\\_(ツ)_/¯");
				})
				.then(() => (
					setState({
						isFetching: false,
						scrollEnd: false
					})
				));

		}

	}, [breed, subBreed, state.scrollEnd, state.limit])

	const renderTitle = (
		<h2 className='mb-20 ez-text'>
			Here are some more doggos like yours
		</h2>
	);

	const renderSpinner = (
		<div className='spinner'>
			<Loading lg secondary />
		</div>
	);

	return state.images.length > 0 ? (
		<>
			{renderTitle}
			<Masonry
				breakpointCols={{
					1100: 3,
					1600: 4,
					2000: 5,
					2500: 6,
					500: 1,
					700: 2,
					default: 3
				}}
				className="gallery"
				columnClassName="gallery-column"
			>
				{
					state.images.map((image, i) => (
						<GalleryItem
							image={image}
							key={i.toString() + breed + subBreed}
						/>
					))
				}
			</Masonry>
			{state.isFetching && renderSpinner}
		</>
	) : (
		state.isFetching ? renderSpinner : props.empty
	);
}

export default Gallery;
