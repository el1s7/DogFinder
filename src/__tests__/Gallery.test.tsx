import { render, screen } from '@testing-library/react';
import React from 'react';
import renderer from 'react-test-renderer';

import Gallery from './../Components/Gallery';

it('gallery empty render correctly', () => {
	expect.hasAssertions();
	render(<Gallery
		breed={['', '']}
		empty={<span>This is empty</span>}
	/>);
	const emptyElement = screen.getByText(/this is empty/ui);
	expect(emptyElement).toBeInTheDocument();
});


it('gallery images render correctly', () => {
	expect.hasAssertions();

	const component = renderer
		.create(
			<Gallery
				breed={['', '']}
				empty={<span>Empty</span>}
				initialImages={[
					'/akita.png',
					'/rottweiler.jpg',
					'/golden.jpg',
					'/prancer.jpg'
				]}
			/>
		)

	const tree = component.toJSON();

	expect(tree).toMatchSnapshot();
});
