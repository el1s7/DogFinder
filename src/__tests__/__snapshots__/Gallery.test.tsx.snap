// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gallery images render correctly 1`] = `
Array [
  <h2
    className="mb-20 ez-text"
  >
    Here are some more doggos like yours
  </h2>,
  <div
    className="gallery"
  >
    <div
      className="gallery-column"
      style={
        Object {
          "width": "33.333333333333336%",
        }
      }
    >
      <div
        className="gallery-item"
      >
        <div
          className="gallery-image"
          onClick={[Function]}
        >
          <img
            alt="A dog"
            loading="lazy"
            src="/akita.png"
          />
        </div>
        <div
          className="preview-container"
          onClick={[Function]}
        >
          <img
            alt="A dog"
            src="/akita.png"
          />
          <div
            className="preview-close"
            onClick={[Function]}
          >
            <i
              className="zwicon-close"
            />
          </div>
        </div>
      </div>
      <div
        className="gallery-item"
      >
        <div
          className="gallery-image"
          onClick={[Function]}
        >
          <img
            alt="A dog"
            loading="lazy"
            src="/prancer.jpg"
          />
        </div>
        <div
          className="preview-container"
          onClick={[Function]}
        >
          <img
            alt="A dog"
            src="/prancer.jpg"
          />
          <div
            className="preview-close"
            onClick={[Function]}
          >
            <i
              className="zwicon-close"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="gallery-column"
      style={
        Object {
          "width": "33.333333333333336%",
        }
      }
    >
      <div
        className="gallery-item"
      >
        <div
          className="gallery-image"
          onClick={[Function]}
        >
          <img
            alt="A dog"
            loading="lazy"
            src="/rottweiler.jpg"
          />
        </div>
        <div
          className="preview-container"
          onClick={[Function]}
        >
          <img
            alt="A dog"
            src="/rottweiler.jpg"
          />
          <div
            className="preview-close"
            onClick={[Function]}
          >
            <i
              className="zwicon-close"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="gallery-column"
      style={
        Object {
          "width": "33.333333333333336%",
        }
      }
    >
      <div
        className="gallery-item"
      >
        <div
          className="gallery-image"
          onClick={[Function]}
        >
          <img
            alt="A dog"
            loading="lazy"
            src="/golden.jpg"
          />
        </div>
        <div
          className="preview-container"
          onClick={[Function]}
        >
          <img
            alt="A dog"
            src="/golden.jpg"
          />
          <div
            className="preview-close"
            onClick={[Function]}
          >
            <i
              className="zwicon-close"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
